src/providers/files/files_ops.c

```diff
@@ -696,101 +684,287 @@ static errno_t sf_enum_groups(struct files_id_ctx *id_ctx,
+            } else if (ret == ENOENT) {
                 DEBUG(SSSDBG_MINOR_FAILURE,
                       "The file %s does not exist (yet), skipping\n",
-                      id_ctx->group_files[i]);
-                continue;
+                      filename);
             } else if (ret != EOK) {
                 DEBUG(SSSDBG_OP_FAILURE,
                       "Cannot enumerate groups from %s, aborting\n",
-                      id_ctx->group_files[i]);
+                      filename);
                 goto done;
             }
+        } else {
+            state->current_step = UPDATE_FINISH;
+        }
+        break;
+    case SAVE_GROUPS:
+        DEBUG(SSSDBG_TRACE_ALL, "Step SAVE_GROUPS.\n");
+        if (state->groups != NULL) {
+            ret = sf_enum_groups(id_ctx, state->groups,
+                                 state->obj_idx, state->batch_size);
+            if (ret == EOK) {
+                state->current_step = READ_GROUPS;
+            } else if (ret == EAGAIN) {
+                state->obj_idx += state->batch_size;
+            } else {
+                DEBUG(SSSDBG_OP_FAILURE, "Saving groups failed.\n");
+                goto done;
+            }
+        }
+        break;
+    case UPDATE_FINISH:
+        DEBUG(SSSDBG_TRACE_ALL, "Step UPDATE_FINISH.\n");
+        ret = dp_add_sr_attribute(id_ctx->be);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Failed to add session recording attribute, ignored.\n");
         }
-    }
 
-    ret = dp_add_sr_attribute(id_ctx->be);
-    if (ret != EOK) {
-        DEBUG(SSSDBG_CRIT_FAILURE,
-              "Failed to add session recording attribute, ignored.\n");
-    }
+        ret = sysdb_transaction_commit(id_ctx->domain->sysdb);
+        if (ret != EOK) {
+            goto done;
+        }
+        state->in_transaction = false;
 
-    ret = sysdb_transaction_commit(id_ctx->domain->sysdb);
-    if (ret != EOK) {
+        state->current_step = UPDATE_DONE;
+
+        ret = EOK;
+        break;
+    default:
+        DEBUG(SSSDBG_CRIT_FAILURE, "Undefined update step [%zu].\n",
+                                   state->current_step);
+        ret = EINVAL;
         goto done;
     }
-    in_transaction = false;
 
-    ret = EOK;
+    if (state->current_step != UPDATE_DONE) {
+        tv = tevent_timeval_current_ofs(0, state->delay);
+        state->te = tevent_add_timer(id_ctx->be->ev, state, tv,
+                                     sf_enum_files_first_step, req);
+        if (state->te == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "Unable to schedule files update.\n");
+            ret = EFAULT;
+            goto done;
+        }
+
+        return;
+    }
+
 done:
-    if (in_transaction) {
+    if (state->in_transaction) {
         tret = sysdb_transaction_cancel(id_ctx->domain->sysdb);
         if (tret != EOK) {
             DEBUG(SSSDBG_CRIT_FAILURE,
                   "Cannot cancel transaction: %d\n", ret);
         }
```

Comment:
Shouldn't we set `state->in_transaction = false` here?

Reply:
Hi,

I added it for consistency, since the request is ending here and the transaction state isn't checked afterwards I think it is not required.

bye,
Sumit